cmd/wstail/tail.go

```diff
@@ -0,0 +1,242 @@
+package main
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io"
+	"net/http"
+	"os/exec"
+	"syscall"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/gorilla/websocket"
+	"github.com/gravitational/trace"
+)
+
+const defaultTailSource = "/var/log/messages"
+
+// tailHistory defines the number of lines to output
+const tailHistory = 100
+
+func main() {
+	log.SetLevel(log.InfoLevel)
+	flag.Parse()
+	if flag.NArg() < 1 {
+		filePath = defaultTailSource
+	} else {
+		filePath = flag.Args()[0]
+	}
+
+	http.Handle("/ws", makeHandler(serveWs))
+
+	log.Infof("listening on %v", *addr)
+	log.Fatalln(http.ListenAndServe(*addr, nil))
+}
+
+func writer(ws *websocket.Conn, filter filter) {
+	matcher := buildMatcher(filter)
+	log.Infof("active filter: %v (%v)", filter, matcher)
+	defer ws.Close()
+
+	tailCmd := exec.Command("tail", fmt.Sprintf("-n%v", tailHistory), "-f", filePath)
+	commands := []*exec.Cmd{tailCmd}
+	if matcher != "" {
+		// --line-buffered is not supported in busybox
+		// grepCmd := exec.Command("grep", "--line-buffered", "-E", matcher)
+		grepCmd := exec.Command("grep", "-E", matcher)
+		commands = append(commands, grepCmd)
+	}
+	pipe, err := pipeCommands(commands...)
+	if err != nil {
+		log.Errorf("failed to build a command pipeline: %v", err)
+	}
+	defer pipe.Close()
+
+	messageC := newMessagePump(pipe)
+	closeNotifierC := newCloseNotifierLoop(ws)
+
+	var errDisconnected error
+	for errDisconnected == nil {
+		select {
+		case <-closeNotifierC:
+			log.Infof("client disconnected")
+			return
+		case line := <-messageC:
+			var payload = struct {
+				Type    string `json:"type"`
+				Payload string `json:"payload"`
+			}{
+				Type:    "data",
+				Payload: string(line),
+			}
+
+			if data, err := json.Marshal(&payload); err != nil {
+				log.Infof("failed to convert to JSON: %v", err)
+			} else {
+				errDisconnected = ws.WriteMessage(websocket.TextMessage, data)
+			}
+		}
+	}
+}
+
+// newCloseNotifierLoop spawns a goroutine that reads from the client.
+// The loop terminates once a message from the client has been received.
+// The only expected message the close message although the goroutine does not validate this fact.
+// Returns a channel that will be closed if client disconnects.
+func newCloseNotifierLoop(ws *websocket.Conn) chan struct{} {
+	notifierC := make(chan struct{})
+	go func() {
+		var err error
+		for err == nil {
+			var msg []byte
+			_, msg, err = ws.ReadMessage()
+			if err == nil {
+				log.Infof("recv: %s", msg)
+			}
+		}
+		log.Infof("closing connection with %v", err)
+		close(notifierC)
+	}()
+	return notifierC
+}
+
+// newMessagePump spawns a goroutine to handle messages from the tailing process group.
+// Returns a channel where the received messages are sent to.
+func newMessagePump(r io.Reader) chan []byte {
+	messageC := make(chan []byte)
+	go func() {
+		s := bufio.NewScanner(r)
```

Comment:
IMHO this is a fine use of `s` since it's not referenced beyond the next handful of lines, just to frame my comments in other places.